cmdundo.go

```diff
@@ -25,37 +30,155 @@ var (
+	drop := make(map[string]bool)
+	for _, r := range modFile.Replace {
+		if !modMap[r.Old.Path] || r.Old.Version != "" || r.New.Version != "" {
+			continue
+		}
+		// Found a replacement to drop.
+		comments := r.Syntax.Comments
+		if len(comments.Suffix) == 0 {
+			// No comment; we can just drop it.
+			drop[r.Old.Path] = true
+			delete(modMap, r.Old.Path)
+			continue
+		}
+		prevReplace := splitWasComment(comments.Suffix[0].Token)
+		if prevReplace != nil && prevReplace.Old.Path == r.Old.Path {
+			// We're popping the old replace statement.
+			if !r.Syntax.InBlock {
+				// When we're not in a block, we don't need the "replace" token.
+				prevReplace.Syntax.Token = prevReplace.Syntax.Token[1:]
+			}
+			// Preserve any before and after comments.
+			prevReplace.Syntax.Before = r.Syntax.Before
+			prevReplace.Syntax.After = r.Syntax.After
+			*r = *prevReplace
+			delete(modMap, r.Old.Path)
+		} else {
+			// It's not a "was" comment. Just remove it (after this loop so we don't
+			// interfere with the current range statement).
+			drop[r.Old.Path] = true
+			delete(modMap, r.Old.Path)
+		}
 	}
-	if _, err := runCmd(cwd, "go", args...); err != nil {
-		return errors.Notef(err, nil, "failed to remove go.mod replacements")
+	for m := range drop {
+		if err := modFile.DropReplace(m, ""); err != nil {
+			return errors.Notef(err, nil, "cannot drop replacement for %v", m)
+		}
+	}
+	failed := make([]string, 0, len(modMap))
+	for m := range modMap {
+		failed = append(failed, m)
+	}
+	sort.Strings(failed)
+	for _, m := range failed {
+		errorf("%s not currently replaced; cannot drop", m)
+	}
+	if err := writeModFile(modFile); err != nil {
+		return errors.Wrap(err)
 	}
 	for _, m := range modules {
 		fmt.Printf("dropped %s\n", m)
 	}
 	return nil
 }
+
+func splitWasComment(s string) *modfile.Replace {
+	s1 := strings.TrimPrefix(s, "// was ")
+	if len(s1) == len(s) {
+		return nil
+	}
+	s = s1
+	i := strings.Index(s, " => ")
+	if i == -1 {
+		return nil
+	}
+	old, ok := splitPathVersion(s[0:i])
+	if !ok {
+		return nil
+	}
+	s = s[i+len(" => "):]
+	i = strings.Index(s, " //")
+	oldComment := ""
+	if i != -1 {
+		oldComment = s[i+1:]
+		s = s[0:i]
+	}
+	new, ok := splitPathVersion(s)
+	if !ok {
+		return nil
+	}
+	r := &modfile.Replace{
+		Old: old,
+		New: new,
+		Syntax: &modfile.Line{
+			Token: tokensForReplace(old, new),
+		},
+	}
+	if oldComment != "" {
+		r.Syntax.Comments.Suffix = []modfile.Comment{{
+			Token:  oldComment,
+			Suffix: true,
+		}}
+	}
+	return r
+}
+
+func tokensForReplace(old, new module.Version) []string {
+	tokens := make([]string, 0, 6)
+	tokens = append(tokens, "replace")
+	tokens = append(tokens, old.Path)
+	if old.Version != "" {
+		tokens = append(tokens, old.Version)
+	}
+	tokens = append(tokens, "=>")
+	tokens = append(tokens, new.Path)
+	if new.Version != "" {
+		tokens = append(tokens, new.Version)
+	}
+	return tokens
+}
+
+func splitPathVersion(s string) (module.Version, bool) {
```

Comment:
nit - why not return just `module.Version`, then check if it's non-nil? Some types, like `go/token.Position`, have an `IsValid() bool` method to make the code easier to write.

Reply:
not sure. I think the bool return is quite clear, and the code not much longer with it. i'll leave it for now.